### ISSUE TITLE
Could you please provide your training code ? Thank you so much

### ISSUE THREAD

**Issue by kehuantiantang** [kkanshul/dock#1]

**Comment by kkanshul** [comment]
Hi kehuantiantang, training code is in Torch Lua which is dead currently. I will suggest implementing DOCK in PyTorch. You can refer the test code which I provided as training will be similar.  After the Line 49 in test.lua use the actual prior scores during training. You can refer for more details.

**Comment by kehuantiantang** [comment]
Thanks a lot for your reply. 
It is very important to  build the classify for those four common-sense (similarity, attribute, spatial, sense) in your paper. If you don't mind, could you please share some code of those classifies or detail common sense matrix description.

For example, spatial common-sense, ![image](https://user-images.githubusercontent.com/5328587/64064380-8485aa80-cc3b-11e9-95b5-e00d1aa55c35.png) encoding the pixel-wise probability of all other source objects under a given relation and<img src="https://latex.codecogs.com/gif.latex?$Y^S_{c_j,&space;rel}$" title="$Y^S_{c_j, rel}$" /> is the relative size distribution.
- What is the different between <img src="https://latex.codecogs.com/gif.latex?$Y^S_{c_j,&space;rel}$" title="$Y^S_{c_j, rel}$" /> and ![image](https://user-images.githubusercontent.com/5328587/64064380-8485aa80-cc3b-11e9-95b5-e00d1aa55c35.png).
- Why the distribution can learn just use the source classes and reused for any of the target classes? Is it means that the probability of every object and relationship is fixed  and just select the max probability one when compute the <img src="https://latex.codecogs.com/gif.latex?$Y_{sp}$" title="$Y_{sp}$" /> matrix ?

**Comment by kkanshul** [comment]
Hi kehuantiantang,

![image](https://user-images.githubusercontent.com/11766853/64204258-3db9cf80-ce4a-11e9-9b80-5339579eda2e.png) capture the distribution at pixel level for class cj and relarion rel. So, basically it is a 2d map relative to the cj position in the image. Higher the value more likely for an object with relation rel with cj to be present at that location. So, for any given proposal you can find its score by seeing value of its center in the 2d map. Where as   ![image](https://user-images.githubusercontent.com/11766853/64204529-cdf81480-ce4a-11e9-994e-586791c4a85e.png) capture the relative size distribution. This captures the distribution of the ratio of cj area and proposal areas for any given realtion rel.  So, for any new proposal you can compute ratio of its area and cj area and get the score using learned distribution(![image](https://user-images.githubusercontent.com/11766853/64205894-f2092500-ce4d-11e9-8e75-ebcf9e41abc3.png))

The learnt distribution is fixed for a class cj (one of the source class) and relation rel. For example, you can learn the distribution of 'table' and 'on'. For learning this distribution you can use other source classes like 'bottle' and 'plant'. Once learned,  the distribution of 'table' and 'on' can be used for other target classes like 'vase', 'spoon', 'fork' etc. Once learned through source classes, these distributions are fixed. Let me know if it is not clear.

**Comment by kehuantiantang** [comment]
Hi, kkanshul,
The  sentence "capture the distribution at **pixel level**" still confuse me.
For example, we can the spatial common-sense relationship from Visual Genome which likes {object1, relation, objects}
Is it means the <img src="https://latex.codecogs.com/gif.latex?\gamma^L_{c_j,&space;rel}" title="\gamma^L_{c_j, rel}" /> generated by count their frequent statistics(<object1, relation>)? If it works like that how to explain the 2d map distribution? 
I am so sorry that I still can not understand how to get the spatial common-sense distribution image in paper Fig6.

**Comment by kkanshul** [comment]
Hi kehuantiantang,

We use frequency statistics but also bounding boxes of the source classes. For <object1, relation, object2>, you will know the bounding boxes of object1 and object2. Once you normalized the boxes, then you can use relative position of object2 with respect to the object1 to create 2d map. For  example, let object1 be table and relation be 'on'. Then you will find bounding boxes of all object2 (object2 should be one of the source class) which is associated with the table with relation 'on'. And then use these boxes to create 2 map (remember these boxes are normalized and are relative to table box). Let me know if this is not clear.

**Comment by kehuantiantang** [comment]
Hi kkanshul
Thank you so much that your explain give me a lot of help.So follow your work, if I want to build ![image](https://user-images.githubusercontent.com/5328587/64616530-b9ea7f00-d417-11e9-9068-f0c016db9515.png) for relationship <table, on>,  the algorithm seems to be
```
1. find all the bounding box for table, and relationship for on, total N bounding box
2. normalization bounding box, then set B_{i,j} = 1
3. initial the 2d relation distribution, Y = 0
4, Y = Y + B_{i, j}
5, normalize Y, Y_{normal} = Y/N
```
Is it right ? 
In order to avoid confusion, I draw a graph to explain algorithm.
![naF8Bt.jpg](https://s2.ax1x.com/2019/09/10/naF8Bt.jpg)

Thanks a lot, ^_^

**Comment by kkanshul** [comment]
It looks fine, but remember you need to draw map relative to table. So, you should place table at same postion in 2d map irrespective of its position in the image (center would be easiest). And then change the position of other bounding boxes relative to table.

**Comment by kehuantiantang** [comment]
Dear kkanshui,
It's clear to me now. Thank you so much for your help.
